plugin/util.go

```diff
@@ -0,0 +1,547 @@
+		logrus.Println("Error getting xml report data: ", err.Error())
+		return tagsMap, fieldsMap, err
+	}
+
+	totalAggregate := calculateAggregate(aggregatorList)
+	logrus.Println("Total Aggregate: ", totalAggregate)
+
+	pipelineId, buildNumber, err := GetPipelineInfo()
+	if err != nil {
+		logrus.Println("Error getting pipeline info: ", err.Error())
+		return tagsMap, fieldsMap, err
+	}
+
+	tagsMap, fieldsMap = getDataMaps(pipelineId, buildNumber, totalAggregate)
+	err = showBuildStats(tagsMap, fieldsMap)
+	if err != nil {
+		logrus.Println("Error showing build stats: ", err.Error())
+		return tagsMap, fieldsMap, err
+	}
+
+	if dbUrl != "" && dbToken != "" && dbOrg != "" && dbBucket != "" {
+		err = PersistToInfluxDb(dbUrl, dbToken, dbOrg, dbBucket, measurementName, groupName, tagsMap, fieldsMap)
+		if err != nil {
+			logrus.Println("Error persisting data to InfluxDB: ", err.Error())
+			return tagsMap, fieldsMap, err
+		}
+	}
+
+	return tagsMap, fieldsMap, err
+}
+
+func PersistToInfluxDb(dbUrl, dbToken, dbOrganisation, dbBucket, measurementName, groupName string,
+	tagsMap map[string]string, fieldsMap map[string]interface{}) error {
+
+	tagsMap["group"] = groupName
+	client := influxdb2.NewClient(dbUrl, dbToken)
+	defer client.Close()
+	writeAPI := client.WriteAPIBlocking(dbOrganisation, dbBucket)
+
+	point := influxdb2.NewPoint(
+		measurementName,
+		tagsMap,
+		fieldsMap,
+		time.Now())
+	err := writeAPI.WritePoint(context.Background(), point)
+	if err != nil {
+		logrus.Println("Error writing point: ", err)
+		return err
+	}
+	logrus.Println("Data persisted successfully to InfluxDB.")
+	return nil
+}
+
+func GetPipelineInfo() (string, string, error) {
+	pipelineId := os.Getenv(PipeLineIdEnvVar)
+	buildNumber := os.Getenv(BuildNumberEnvVar)
+
+	if pipelineId == "" || buildNumber == "" {
+		return "", "", fmt.Errorf("PipelineId or BuildNumber not found in the environment")
+	}
+
+	return pipelineId, buildNumber, nil
+}
+
+func GetPreviousBuildId(measurementName, influxURL, token, org, bucket, currentPipelineId, groupId, currentBuildId string) (prevBuildId int, err error) {
+	client := influxdb2.NewClient(influxURL, token)
+	defer client.Close()
+
+	query := fmt.Sprintf(`
+	from(bucket: "%s")
+	  |> range(start: -1y)
+	  |> filter(fn: (r) => r._measurement == "%s")
+	  |> filter(fn: (r) => r.pipelineId == "%s")
+	  |> filter(fn: (r) => r.group == "%s")
+	  |> keep(columns: ["buildId"])
+	  |> distinct(column: "buildId")
+	  |> sort(columns: ["buildId"], desc: true)
+	`, bucket, measurementName, currentPipelineId, groupId)
+
+	queryAPI := client.QueryAPI(org)
+	result, err := queryAPI.Query(context.Background(), query)
+	if err != nil {
+		return 0, fmt.Errorf("failed to query InfluxDB: %w", err)
+	}
+
+	var buildIds []int
+	for result.Next() {
+		buildIdStr := result.Record().ValueByKey("buildId")
+		if buildIdStr == nil {
+			continue
+		}
+		buildId, convErr := strconv.Atoi(buildIdStr.(string))
+		if convErr != nil {
+			continue
```

Comment:
add a log here `Error converting buildId '%s' to integer: %v", buildIdStr, convErr`